Added create record functionality

AnastasiaShemyakinskaya opened the pull request:
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Added create record functionality
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Fixed mistakes
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

paulgrape commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -27,17 +61,31 @@ public KafkaStorageDriver(
   protected final boolean submit(final O op) throws IllegalStateException {
     if (concurrencyThrottle.tryAcquire()) {
       final var opType = op.type();
-      if (opType.equals(OpType.NOOP)) {
```
Why did you remove NOOP check?

AnastasiaShemyakinskaya replied:
It's my mistake, I will fix it, thanks.

evtimo commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -12,13 +14,45 @@
+  private final AtomicInteger rrc = new AtomicInteger(0);
+  private final Map<Properties, AdminClientCreateFunctionImpl> adminClientCreateFuncCache =
+      new ConcurrentHashMap<>();
+  private final Map<String, AdminClient> adminClientCache = new ConcurrentHashMap<>();
```
As I understood, AdminClient doesn't need caching, it can be like final variable in constructor.

akurilov replied:
@zhenytim AdminClient depends on the selected Kafka node address, so it should be cached.

AnastasiaShemyakinskaya committed:
Restore NOOP operation

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -12,13 +14,45 @@
 
+  private final String[] endpointAddrs;
+  private final int nodePort;
+  private boolean key;
```
Please use final for the fields initialized in the constructor

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -27,17 +61,34 @@ public KafkaStorageDriver(
   protected final boolean submit(final O op) throws IllegalStateException {
     if (concurrencyThrottle.tryAcquire()) {
       final var opType = op.type();
+      var nodeAddr = op.nodeAddr();
```
final

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
 
-  private void submitRecordCreateOperation() {}
+  boolean completeFailedOperation(final O op, final Throwable thrown) {
+    thrown.printStackTrace();
```
Please use the logger instead:
`LogUtil.exception(Level.DEBUG, e, "{}: operation failed: {}", stepId, op);`

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
+      var kafkaProducer = producerCache.computeIfAbsent(nodeAddr, producerConfig);
+      var recordItem = recordOp.item();
+      var topicName = recordItem.name();
+      if (topicCache.get(topicName) == null) {
```
The approach is not thread safe. Please use the method `computeIfAbsent` of `ConcurrentMap` interface to check and instantiate (if necessary) atomically. Also use some method reference instead of the constructor

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
+        topicCache.put(topicName, topic);
+        adminClient.createTopics(Collections.singletonList(topic));
+      }
+      Callback callback =
```
Please use the method reference instead of callback. The driver class can implement the Callback interface for this either the callback may be a field of the driver. This is required because we don't want to create new callback instance every time.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
+            }
+          };
+      if (key) {
+        var producerKey = Long.toString(recordItem.offset(), Character.MAX_RADIX);
```
Let's use item.name() - it would be much more simple

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -12,13 +14,45 @@
+  private boolean key;
+  private int numPartition;
+  private short replicationFactor;
+  private int request;
```
rename to something like "requestSizeLimit"

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -71,7 +122,88 @@ private void submitRecordDeleteOperation() {}
+
+  private void submitRecordCreateOperation(final DataOperation recordOp, String nodeAddr) {
+    try {
+      var adminConfig =
```
1. Let's use the config cache in order to not create the config on each invocation
2. You can use single producer config instead of admin config too
3. Use `val` instead of `var` for the immutable values. This brings in a bit more safety

AnastasiaShemyakinskaya committed:
Merge branch 'master' of https://github.com/emc-mongoose/mongoose-storage-driver-kafka into KAFKA-33-record-functionality

AnastasiaShemyakinskaya committed:
Fixed wrong merge
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Resolve comments
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
added lambda expressions
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Fixed compilation error
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Fixed error with exception
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -75,7 +127,85 @@ private void submitRecordDeleteOperation() {}
+          producerCreateFuncCache.computeIfAbsent(config, ProducerCreateFunctionImpl::new);
+      val kafkaProducer = producerCache.computeIfAbsent(nodeAddr, producerConfig);
+      val recordItem = recordOp.item();
+      val topicName = recordItem.name();
```
The record item name will look like `/<somepath>/<somename>`. I suggest to use the `<somepath>` for the topic name and `<somename>` for the key. Example:
https://github.com/emc-mongoose/mongoose-storage-driver-pravega/blob/1f693952f82dfe86ff52f5bea7587d8816c46902/src/main/java/com/emc/mongoose/storage/driver/pravega/PravegaStorageDriver.java#L467

AnastasiaShemyakinskaya replied:
As I understand topicName = recordOp.dstPath(), key = recordItem,name()

akurilov replied:
Yes, if dstPath doesn't include item name (I'm unsure)

AnastasiaShemyakinskaya committed:
Change topicName, fixed exception check in callback
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com

AnastasiaShemyakinskaya committed:
Merge branch 'master' into KAFKA-33-record-functionality

AnastasiaShemyakinskaya committed:
Removed unnecessary lines
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

Signed-off-by: Anastasia Shemyakinskaya shem98a@gmail.com
